src/core/lib/iomgr/tcp_uv.cc

```diff
@@ -218,26 +243,52 @@ static grpc_error* uv_socket_init(grpc_custom_socket* socket, int domain) {
   if (error != GRPC_ERROR_NONE) {
     return error;
   }
-  uv_socket->handle->data = socket;
+  ((uv_handle_t*)uv_socket->handle.ptr)->data = socket;
   socket->impl = uv_socket;
   return GRPC_ERROR_NONE;
 }
 
 static grpc_error* uv_socket_getpeername(grpc_custom_socket* socket,
                                          const grpc_sockaddr* addr,
                                          int* addr_len) {
+  int err;
   uv_socket_t* uv_socket = (uv_socket_t*)socket->impl;
-  int err = uv_tcp_getpeername(uv_socket->handle,
-                               (struct sockaddr*)IGNORE_CONST(addr), addr_len);
+  if (uv_socket->handle_type == HANDLE_TYPE_PIPE) {
+#ifdef GRPC_HAVE_UNIX_SOCKET
+    struct sockaddr_un* un_addr = (struct sockaddr_un*)addr;
+    size_t len = sizeof(un_addr->sun_path) - 1;
+    err = uv_pipe_getpeername(uv_socket->handle.pipe, un_addr->sun_path, &len);
+    *addr_len = len + sizeof(un_addr->sun_family) + 1;
+    un_addr->sun_family = AF_UNIX;
+#else
+    err = ENOTSUP;
+#endif
+  } else {
+    err = uv_tcp_getpeername(uv_socket->handle.tcp,
+                             (struct sockaddr*)IGNORE_CONST(addr), addr_len);
+  }
   return tcp_error_create("getpeername failed", err);
 }
 
 static grpc_error* uv_socket_getsockname(grpc_custom_socket* socket,
                                          const grpc_sockaddr* addr,
                                          int* addr_len) {
+  int err;
   uv_socket_t* uv_socket = (uv_socket_t*)socket->impl;
-  int err = uv_tcp_getsockname(uv_socket->handle,
-                               (struct sockaddr*)IGNORE_CONST(addr), addr_len);
+  if (uv_socket->handle_type == HANDLE_TYPE_PIPE) {
```

Comment:
Maybe split this out to a helper function. You're already having a duplicate of this a few lines above, and the readability is seriously decreasing due to the added `#ifdef` in the middle.

Please consider defining two helper functions that are mutually exclusive with an `#ifdef` around them.